### ISSUE TITLE
close code 1006 ("no reason") even when reason should be known

### ISSUE THREAD

**Issue by cjerdonek** [aaugustin/websockets#877]
This issue is about connections showing up as "closed" in the logs with code 1006 ("no reason") even though they started out as "closing" with a particular reason (i.e. non-1006, for example 1009). The reason this is an issue is that if the final code reflected the initial reason, then error closures would be easier for users to diagnose. Currently, this isn't especially easy (see [this question from the FAQ](https://websockets.readthedocs.io/en/stable/faq.html#what-does-connectionclosederror-code-1006-mean), for instance).

Details:

When troubleshooting a particular connection closure on the server side with code 1006, it was a little surprising for me to see that the server initiated the closure with code 1009 ("message too big"). In other words, if the reason is 1009 ("message too big"), then why is the reason showing up as "no reason"?

Here is a sample snippet from websockets' logs on the server side:

```
DEBUG:websockets.protocol:server - event = data_received(<1963 bytes>)
DEBUG:websockets.protocol:server ! failing OPEN WebSocket connection with code 1009
DEBUG:websockets.protocol:server - state = CLOSING
DEBUG:websockets.protocol:server > Frame(fin=True, opcode=8, data=b'\x03\xf1', rsv1=False, rsv2=False, rsv3=False)
DEBUG:websockets.protocol:server x half-closing TCP connection
DEBUG:websockets.protocol:server - event = data_received(<8 bytes>)
DEBUG:websockets.protocol:server - event = eof_received()
DEBUG:websockets.protocol:server - event = connection_lost(None)
DEBUG:websockets.protocol:server - state = CLOSED
DEBUG:websockets.protocol:server x code = 1006, reason = [no reason]
```

When I started looking into why 1009 isn't reflected in the final code, I came across [this snippet in the spec](https://tools.ietf.org/html/rfc6455#section-5.5.1):

> If an endpoint receives a Close frame and did not previously send a Close frame, the endpoint MUST send a Close frame in response. (When sending a Close frame in response, the endpoint typically echos the status code it received.) It SHOULD do so as soon as practical. An endpoint MAY delay sending a Close frame until its current message is sent (for instance, if the majority of a fragmented message is already sent, an endpoint MAY send the remaining fragments before sending a Close frame). However, there is no guarantee that the endpoint that has already sent a Close frame will continue to process data.

The above log snippet was generated by running websockets on both the client and server side, so this made me wonder: is websockets coded to echo the status code it received? (This is the part about, "When sending a Close frame in response, the endpoint typically echos the status code it received. It SHOULD do so as soon as practical.") Because of the logs above, it seems like it doesn't. If it did, it seems like the reason would be reflected in the final code instead of "no reason," which it seems would be an improvement in ergonomics.

**Comment by cjerdonek** [comment]
I also wonder if websockets should be reporting e.g. 1009 instead of 1006 in the case above, regardless of whether the endpoint responded to the close frame with another close frame. One reason is that [the spec's definition of 1006](https://tools.ietf.org/html/rfc6455#section-7.4.1) is--

> 1006 is a reserved value and MUST NOT be set as a status code in a Close control frame by an endpoint. It is designated for use in applications expecting a status code to indicate that the connection was closed abnormally, e.g., without sending or receiving a Close control frame.

In the case where the server initiates the close with a code of 1009, at that point it's no longer expecting a status code in return (i.e. because it _has_ already sent a close control frame, in contrast with the e.g. at the end of the sentence above.)

**Comment by aaugustin** [comment]
Yes, websockets is less than helpful here :-(

Per https://tools.ietf.org/html/rfc6455#section-7.1.5:

> _The WebSocket Connection Close Code_ is
   defined as the status code (Section 7.4) contained in the first Close
   control frame received by the application implementing this protocol.
   If this Close control frame contains no status code, _The WebSocket
   Connection Close Code_ is considered to be 1005.

This is why the API for getting the close code reports 1006 rather than 1009 in your case.

Clearly websockets should provide information about the close code _sent_ in addition to the close code _received_. I'm not sure how.

**Comment by cjerdonek** [comment]
Thanks a lot for entertaining this issue. It may be helpful to discuss two questions I raised separately.

First, there is the question of what websockets does in response to receiving a close frame. Does it echo back a close frame in response with the same status code, or does it terminate the connection without doing so? The first excerpt I quoted above suggests it should echo back the same close code. If websockets did that, t seems that would solve the issue at least in the case where both sides are using websockets (because it would be getting back the same code it sent).

Secondly, there is the question of what websockets should do and report to the user if it initiates the closure with a particular reason and status code, but it doesn't receive a close frame (or status code) in response.

It seems like there may be some ambiguity or perhaps contradiction in the spec on this. For example, what do you make of the part I quoted above from the definition of 1006? It says, "[1006] is designated for use in applications expecting a status code to indicate that the connection was closed abnormally, e.g., _without sending or receiving a Close control frame_." In the example I provided, since a close control frame _was_ sent, it seems like 1006 might not be appropriate. The definition suggests that 1006 is designed for when a close control frame isn't sent _or_ received. Another way to look at this is that the closure isn't abnormal if the server initiated it (i.e. it's closing as expected).

In terms of how to solve this, I do see that when the `ConnectionClosedError` is raised, it already does have access to the information of what initiated the closure because this is visible earlier in the traceback:

```
Traceback (most recent call last):
  File "/.../site-packages/websockets/protocol.py", line 827, in transfer_data
    message = await self.read_message()
  File "/.../site-packages/websockets/protocol.py", line 895, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "/.../site-packages/websockets/protocol.py", line 971, in read_data_frame
    frame = await self.read_frame(max_size)
  File "/.../site-packages/websockets/protocol.py", line 1051, in read_frame
    extensions=self.extensions,
  File "/.../site-packages/websockets/framing.py", line 142, in read
    frame = extension.decode(frame, max_size=max_size)
  File "/.../site-packages/websockets/extensions/permessage_deflate.py", line 132, in decode
    f"Uncompressed payload length exceeds size limit (? > {max_size} bytes)"
websockets.exceptions.PayloadTooBig: Uncompressed payload length exceeds size limit (? > 1048576 bytes)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/.../site-packages/websockets/server.py", line 191, in handler
    await self.ws_handler(self, path)
  File "server.py", line 6, in echo
    async for message in websocket:
  File "/.../site-packages/websockets/protocol.py", line 439, in __aiter__
    yield await self.recv()
  File "/.../site-packages/websockets/protocol.py", line 509, in recv
    await self.ensure_open()
  File "/.../site-packages/websockets/protocol.py", line 812, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
```

But it's not reflected in the tip of the traceback.

Regarding what you said:

> Clearly websockets should provide information about the close code sent in addition to the close code received. I'm not sure how.

There is also this part:

> NOTE: Two endpoints may not agree on the value of _The WebSocket Connection Close Code_. As an example, if the remote endpoint sent a Close frame but the local application has not yet read the data containing the Close frame from its socket's receive buffer, and the local application independently decided to close the connection and send a Close frame, both endpoints will have sent and received a Close frame and will not send further Close frames. Each endpoint will see the status code sent by the other end as _The WebSocket Connection Close Code_. As such, it is possible that the two endpoints may not agree on the value of _The WebSocket Connection Close Code_ in the case that both endpoints _Start the WebSocket Closing Handshake_ independently and at roughly the same time.

Since there are cases where both values may be important, maybe it would make sense to define an exception class that stores and displays both values.

**Comment by aaugustin** [comment]
This is a duplicate of #587. Let's consolidate the discussion over there.

**Comment by cjerdonek** [comment]
@aaugustin [In one of my comments above](https://github.com/aaugustin/websockets/issues/877#issuecomment-752766044) I mentioned that this issue actually raises two distinct issues. Issue #587 is the second one, but the first one is different. Can we keep an issue open to discuss the first, which is independent of #587?

**Comment by aaugustin** [comment]
> First, there is the question of what websockets does in response to receiving a close frame. Does it echo back a close frame in response with the same status code

Yes, that's what websockets does in general. Also it's the default behavior mandated by the spec.

> or does it terminate the connection without doing so?

This happens in cases where websockets Fails the Connection, a concept defined by the spec and implemented in the `fail_connection` method. In that case, websockets sends a close frame in fire'n'forget mode and closes the TCP connection immediately.

> If websockets did that, it seems that would solve the issue at least in the case where both sides are using websockets (because it would be getting back the same code it sent).

All implementations are required to echo the same code so it doesn't matter which software the other side is running.

However, in the case we're discussing here, websockets cannot get back the code it sent. Closing the conection is the sensible behavior.

Indeed, the size limit is for security purposes. If websockets reads the header of a frame that says "hey this frames is 100GB long", it isn't going to chug through the 100GB of data just to get an hypothetical close frame afterwards. So it sends a close frame with code 1009, closes the connection, never sees a response to the close frame, and therefore the close code (per the spec is 1006).

I believe this addresses your first point. Let me know if that's unclear!

**Comment by cjerdonek** [comment]
> This happens in cases where websockets Fails the Connection, a concept defined by the spec and implemented in the fail_connection method. In that case, websockets sends a close frame in fire'n'forget mode and closes the TCP connection immediately.

Thanks. Yes, that clears things up. To complete this point, that portion of the spec says--

> 7.1.7. **Fail the WebSocket Connection** Certain algorithms and specifications require an endpoint to _Fail the WebSocket Connection_.  To do so, the client MUST _Close the WebSocket Connection_, and MAY report the problem to the user (which would be especially useful for developers) in an appropriate manner.

This seems to lend support to my suggestion [here](https://github.com/aaugustin/websockets/issues/587#issuecomment-854224575) and in the following comment, which is to report to the _developer_ the code that was _sent_, as opposed to what is received (which won't be anything since it's fire and forget.) The code that is sent is the information that would be "especially useful for developers" and thus more of an "appropriate manner."

**Comment by aaugustin** [comment]
I agree!

**Comment by aaugustin** [comment]
It looks like we re-converged to "let's fix #587" so I'm going to close this one again. Hopefully I didn't miss anything.

**Comment by gokturkDev** [comment]
just doing a brain dump here after a whole day of debugging in case it helps anyone. In mobile dev, in certain devices after the device exits the airplane mode, I believe there is a short period of time where it uses mobile data instead of wifi (probably cuz its faster to establish). However after a certain time, the os kills it, resulting in this error. So, if you are in mobile dev and wracking your brain over why your connection drops after sometime after exiting airplane mode, try listening to the network changes and reconnect everytime it changes
